design doc: MIR typechecking using representation types

mgree opened the pull request:
### Motivation

- This PR adds a feature that has not yet been specified.

Implicit casts (`varchar_to_text`) can impede join planning; by having MIR use a type system that corresponds to the representation we'll actually use, we can type-safely elide implicit casts.

benesch commented on doc/developer/design/20240522_mir_representation_types.md:
```diff
@@ -0,0 +1,133 @@
+- (`remix++`) Automatically wrap every use of a `varchar` with `varchar_to_text`.
+
+The `eq-type` and `eq-indx` approach came up in early discussions.
+These are fundamentally hacks, with the upside of being quick fixes.
+The downside is the addition of a tricky corner case, some risk (what if keeping the types separate matters somewhere we've forgotten about?), and some technical debt/mess-making.
```
What makes them fundamentally hacks/tech debt? `eq-indx` in particular seems like a narrow case of a smart index selector (let's call it `conv-indx`) that can handle mixed type lookups, e.g. looking up an i32 in an i64 index or vice-versa, and that seems useful/powerful/desirable?

mgree replied:
The hacky/tech-debt/messy feeling here is that it's a very specialized way to work with `ScalarExpr` in a very narrow context, and I worry about duplicated work and inconsistent handling.

But your point about numeric indices makes it clear to me that index selection will need to be a _little bit_ special no matter what---and index creation, too (#27030).

So, I'm coming around to `eq-indx`/`conv-indx`. An advantage of `mirtype` is that it will let us remove many infallible casts wholesale while simplifying our view of types at the MIR level, which will address our immediate problem. But we'll need something like `conv-indx` in general. (With `conv-indx`, you could avoid doing the work of #15476. Or, alternatively, it may be possible to implement `conv-indx` using the techniques from #15476.)

mgree replied:
Something that's been nagging at me, but I may simply be confused about: the hash functions for `i32` and `i64` are not going to be the same, i.e., sign extending `n::i32` to `signext(n)::i64` will likely not hash the same, so I think we'd need some careful logic here. I remember bringing something similar up to @frankmcsherry (I think using an index on `n::i64` to lookup `n+1` or some similarly invertible operation) and him pointing out why it wouldn't work, but I can't remember what the exact issue was.

ggevay replied:
I'm not really sure how `eq-indx` would work in a join when the actual representations are different. If we have an index on `t1` with a `Datum::Int32` key, and an index on `t2` with a `Datum::Int64` key, then at what moment could we do the cast? I think our current join rendering just does exact equality on the arrangement keys as they are, and there is currently no way to give it a closure or something which would sneakily convert between `Int32` and `Int64`.

So, in addition to modifying the index selection in `JoinImplementation`, we'd need to somehow also modify the join rendering (for Differential join it's `mz_join_core.rs`, and for Delta join it's `dogsdogsdogs::operators::half_join::half_join_internal_unsafe`, in the DD repository). This should involve a solution to the hashing problem mentioned above, besides changing how the keys are equated (somewhere around the `seek_key` calls).

(Note that this problem didn't occur in https://github.com/MaterializeInc/materialize/pull/15476, because there it's only one side of the equality that is doing an index lookup, so there we just move the fiddling with the types to the other side of the equality.)

ggevay replied:
We discussed this a bit with @mgree, and the hashing issue might be worse than I thought. We are doing partitioned joins, i.e., both of the arrangements are partitioned by a hash, and pairs of rows that the join will emit need to be in the same partition. But now with a cast on one of the keys, the partitioning would need to be modified on one of the arrangements, and I'm not sure if there is any better way to do this than to rebuild the arrangement, which would completely defeat the purpose.

mgree replied:
Maybe @frankmcsherry and I can talk about how `eq-indx` would work some time this month? I agree that it seems complicated. 😔

ggevay replied:
(Bw., not an ideal solution, but we could add a notice that would make it easier for users to find these implicit cast issues and manually fix them by tweaking their index keys to include the cast: https://github.com/MaterializeInc/materialize/issues/22133)

ggevay replied:
Oh, look what I found!
https://github.com/MaterializeInc/materialize/issues/10871
It's a really well-preserved archeological specimen: haven't been touched or seen in years, and even features the old EXPLAIN output format!

On a more serious note, suddenly finding this long-forgotten, really detailed discussion of these same problems that we are discussing now really demonstrates the cost of the churn of optimizer people...

mgree replied:
Interesting!

One possibility for `eq-idx` is to automatically coerce to "widest-possible" when doing hashing and ordering. That is, we would hash and order `int16`, `int32`, and so on all as if it were `int128` (and we would similarly hash and order `uint16`, `uint32`, and on on as if it were `uint128`). I don't think there's any storage cost to doing this, just a small computational cost. And then we'd be able to automatically use a wider index for narrower values.

We could even treat everything as `uint128` for hashing purposes, but it probably isn't safe to treat signed ints as `uint128` for ordering purposes (if we ever use ordering to express "real" ordering and not arbitrary ordering).

frankmcsherry replied:
Fwiw, apropos the preceding comment, I'm 99% sure that we hash the row that is the key, and have no deeper understanding of whether the columns are which flavors of integer. The problem would "solve itself" if we asked the rendering layer to use 64 bit integers (or whatever our maximum width is) of the right sign. I don't think we want to be in the business of having the rendering layer improv a different hash function based on its guess about which types should be the same (and then similarly a modified `Ord` to order and equate variants correctly, even with different byte representations).

benesch commented on doc/developer/design/20240522_mir_representation_types.md:
```diff
@@ -0,0 +1,133 @@
+Such rewriting feels a bit risky/annoying/confusing, and I'm not sure how it would interact with sources.
+
+## Open questions
+
+How should we report `ReprScalarType`s to the user in `EXPLAIN PLAN`s?
+There's a potential for confusion if we report something as type `string` that the user expected to have type `VARCHAR(18)` or something similar.
+Apparently we already treat `string` as a synonym for `text`... can we somehow choose names that will be unambiguous?
+Say, `mzstring` or `charseq` or `utf8`?
```
Seems fine to just report them in whatever representation is convenient for us. `MirScalarExpr`s aren't SQL expressions, and are hard to map to SQL expressions; I don't think we need to feel any obligation for the type information that annotates `MirScalarExpr`s to map directly to SQL types, either.

Maybe just stick an `r` prefix in front of it to make it clear(er) it's a lower-level "repr" type? `rstring`, `rint32`, etc.

benesch commented on doc/developer/design/20240522_mir_representation_types.md:
```diff
@@ -0,0 +1,133 @@
+
+Having MIR be typechecked in terms of representation types demands a few changes, which can be broken into three PRs:
+
+1. Rename the existing `ScalarType` to `SqlScalarType`. (We may want to rename `ColumnType` and `RelationType`, as well.)
+2. Have MIR work with `ReprScalarType` (see [MVP](#minimal-viable-prototype)).
```
I don't love that we have two parallel naming schemes: "HIR", "MIR", and "LIR" for expressions, and "Sql" and "Repr" for types. Is it right that HIR types are always `SqlScalarTypes` and MIR and LIR types are always `ReprScalarTypes`? Is there some intuition that suggests we'll only need the two ends of the spectrum for types, even though we needed three IRs for expressions? Or do we think it's likely we'll one day have `HirScalarType`, `MirScalarType`, and `LirScalarType`?

aalexandrov replied:
I think we might also call them `HirScalarType` and `MirScalarType`. LIR already has a precedent of using `MirScalarExpr` for its scalars. If we want to have the IR families aligned in their prefixes we might introduce type aliases

```rust
type LirScalarType = MirScalarType;
type LirScalarExpr = MirScalarExpr;
```

mgree replied:
I'm easy on names. Right now LIR is effectively untyped. If we were to add types to LIR, they would be the same as MIR types. (@frankmcsherry, @antiguru, and I have discussed the idea of a bytecode/instruction LIR, which might end up having a finer grained type system.)

benesch reviewed (commented):
Dropped in some comments! My biggest question is around whether we're overindexing on textual implicit casts and underindexing on implicit casts between integers of different widths. There may be other benefits to representation types beyond implicit casts that make this work worth doing even though it doesn't fix all implicit cast problems, but it seemed worth asking the question.

mgree commented on doc/developer/design/20240522_mir_representation_types.md:
```diff
@@ -0,0 +1,133 @@
+
+These implicit coercions---often of the form `varchar_to_text(#c)` for some column number `c`---get in the way during join planning.
+Even if we have an index on column `c` as a `varchar`, we cannot use that index to plan a join on `varchar_to_text(#c)`: not only are the types different, the terms are different, too.
+We know these implicit coercions interfere with joins; they may also interfere with other aspects of optimization.
```
There are two related ideas here:

 1. what types characterize integers?
 2. how do we deal with casts and their possibly-erroring inverses?

## What types characterize integers?

My primary goal was to have the types at the MIR level characterize the `Datum`s the operator would deal with. All of the various string types will just be `Datum::String` at runtime, so one type works for them. But we actually have different numeric `Datum` constructors: many different sizes of ints (in unsigned and signed flavors),  two sizes of floats, and arbitrary precision numerics. So If MIR types characterize `Datum` constructors, then it seems we would probably need different numeric types.

I can imagine a system where there's a unified numeric supertype `Num` which subsumes all of these individual types, i.e., a value of type num could be any possible numeric `Datum`. But a secondary goal is to avoid complex/automatic typing relationships; one proxy for this complexity is how much overlap there is in the mapping from MIR types to `Datum` constructors. It's unfortunately not possible to have an injective mapping from MIR types to `Datum` constructors, because `Datum::True` and `Datum::False` are booleans but also JSON values... but adding a `Num` supertype would be fairly complicating. And, it wouldn't resolve our other issue...

## How do we deal with casts and their possibly-erroring inverses?

Most of our conversions have inverses---`text_to_varchar` has `varchar_to_text`, timestamps can be converted to and from strings, and the numeric types can move into and out of each other. But in general, these are not true inverses/bijections: one side might produce an error.

⚠️ **Unnecessary math jargon**: ⚠️ `text_to_varchar` is a left inverse of `varchar_to_text`, but not a right inverse. That is, `text_to_varchar[k](varchar_to_text[k](s)) == s` for all `s` of type `varchar(k)`, for all `k`. But it is not the case that `varchar_to_text[k](text_to_varchar[k](s))` for all `s` of type `text` for all `k`---it could be that `s` is longer than `k` and `text_to_varchar[k](s)` produces an error. Put differently, `varchar_to_text` is a _section_ and `text_to_varchar` is a _retraction_.

In general, there's no way to remove the fallible casts and adding in a catch-all numeric type without introducing something worse: pervasive checks and coercions between numeric types.

I think the right way forward here is:

1. Use `UnaryFunc::could_error()` with `UnaryFunc::inverse()`---if neither can error, we can simplify `f(f_inv(x))` to `x`.
2. Identify left-inverses explicitly: if we have `f(f_inv(x))` and only the outer one is fallible, we have a left inverse and we can simplify it to `x`.
3. Explore optimizations around numeric sizing and index selection.

I'm starting to see your point about index selection being a key moment; I'll address it in my response to your other comment.

ggevay replied:
When you say "Identify left-inverses explicitly", you mean that we'd add a new function property, i.e., it's not enough to simply call `could_error` on the inner one, right? Because, in theory, the outer one might fail for some other reason besides not being able to represent the value in the original type. E.g., one might imagine a (thoroughly contrived) function `varchar_to_text_if_not_empty`, which errors if the varchar is empty. This function would be an `inverse` of `text_to_varchar` in the sense that the current code is using the word `inverse`, which is something like "if both `preserves_uniqueness` then it's an inverse for such values for which neither errors".

But maybe we could redefine our `inverse` to mean that you just have to check `could_error` on the inner one. Probably the above contrived case is not actually occurring for any of our functions.

Btw. I just found [an old Slack thread](https://materializeinc.slack.com/archives/C0246GEHL8N/p1692997054843739?thread_ts=1692994729.626079&cid=C0246GEHL8N) where I wrote more details on what the current code means by `inverse` than what to code comment is saying. (I had it on my todo list to actually improve the code comment with this information, but somehow it got buried.)

mgree replied:
You're absolutely right that if we had a function with a declared inverse that errored for some other reason we would have a problem with that approach.

I agree that we should rethink/refine how we think about inverses. Having a single function that says "this is my inverse and here's whether or not I could error" would simplify the thinking here.

mgree commented:
> My biggest question is around whether we're overindexing on textual implicit casts and underindexing on implicit casts between integers of different widths. There may be other benefits to representation types beyond implicit casts that make this work worth doing even though it doesn't fix all implicit cast problems, but it seemed worth asking the question.

You're right that this proposed change really only addresses issues with text data. We represent textual data the same for all pg "character types", so changing the MIR type system lets us conflate them (possibly keeping some cast operations around for their errors).

For numbers, the representations are all different, so this change wouldn't affect us much. I've updated the document a bit, but I guess I'd propose a multi-pronged approach here:

1. Change the type system to deal with textual stuff wholesale.
2. Improve handling of inverses across all types.
3. Implement some form of `conv-indx` across all applicable types/operations.
4. Add notices for "unnecessary" casts: user-written casts that eventually need to be converted back to their original type. (I've seen this pattern in at least two clients, both times in what I would guess was generated code.)

We could do these steps in pretty much any order. I don't have a clear sense of which will have the highest impact.

benesch commented on doc/developer/design/20240522_mir_representation_types.md:
```diff
@@ -87,23 +89,27 @@ pub enum ReprScalarType {
+
+We'll need to define a projection from `HirScalarType` to `MirScalarType` (see [`ScalarType::is_instance_of()`](https://github.com/MaterializeInc/materialize/blob/e4578164fb28a204b43c58ab8ff6c1d3e3b4427f/src/repr/src/scalar.rs#L947) for the correspondence).
+
+In `EXPLAIN PLAN`s, we will report `MirScalarType`s with `r`-prefixed names, where `r` is short for "representation", e.g., `rstring` is the `MirScalarType` corresponding to `text`, `varchar`, etc.
```
Maybe this prefix should be `m` or `mir` if we're going with `MirScalarType`?

mgree replied:
I'm not sure the word `mir` is shown anywhere to users in our current `EXPLAIN` output. Very happy to be told what to do here (the `r` came from [earlier discussion](https://github.com/MaterializeInc/materialize/pull/27239#discussion_r1610903698)).

benesch replied:
I'd go with `mir` (e.g., `mirstring`) to start. I think it's the clearest. Since `EXPLAIN` output isn't stable we have a ton of flexibility to change it if we get user complaints.

benesch commented:
> > My biggest question is around whether we're overindexing on textual implicit casts and underindexing on implicit casts between integers of different widths. There may be other benefits to representation types beyond implicit casts that make this work worth doing even though it doesn't fix all implicit cast problems, but it seemed worth asking the question.
> 
> You're right that this proposed change really only addresses issues with text data. We represent textual data the same for all pg "character types", so changing the MIR type system lets us conflate them (possibly keeping some cast operations around for their errors).
> 
> For numbers, the representations are all different, so this change wouldn't affect us much. I've updated the document a bit, but I guess I'd propose a multi-pronged approach here...
>
> We could do these steps in pretty much any order. I don't have a clear sense of which will have the highest impact.

This makes sense to me! It all seems plausibly quite valuable and as long as we feel good about having `MirScalarType` _and_ `conv-indx` in the limit, no concerns for me. I too don't have a clear sense of which steps will have the highest impact.

> 4\. Add notices for "unnecessary" casts: user-written casts that eventually need to be converted back to their original type. (I've seen this pattern in at least two clients, both times in what I would guess was generated code.)

This one I'm not sold on! In the broad universe of SQL, I don't think there's anything more correct or "natural" about `text` or `varchar`. AFAICT, which you prefer just depends on what database system you were brought up on. My long term ideal UX here is that Materialize absorbs all the complexity here by making both `text` / `varchar` equally efficient, instead of pushing the complexity back up to users and asking them to rewrite their queries (or adjust the application generating queries, which might be ~impossible).

That said, no qualms about doing this first as a stopgap solution, if it's substantially easier than making `text` and `varchar` equally efficient, and then later removing the notice when we've closed the efficiency gap.

mgree commented:
> > 4. Add notices for "unnecessary" casts: user-written casts that eventually need to be converted back to their original type. (I've seen this pattern in at least two clients, both times in what I would guess was generated code.)
> 
> This one I'm not sold on! In the broad universe of SQL, I don't think there's anything more correct or "natural" about `text` or `varchar`. AFAICT, which you prefer just depends on what database system you were brought up on. My long term ideal UX here is that Materialize absorbs all the complexity here by making both `text` / `varchar` equally efficient, instead of pushing the complexity back up to users and asking them to rewrite their queries (or adjust the application generating queries, which might be ~impossible).

That's a good point. Especially considering that many of the casts we're seeing are from generated code, a notice would be annoying (and devalue notices).

> That said, no qualms about doing this first as a stopgap solution, if it's substantially easier than making `text` and `varchar` equally efficient, and then later removing the notice when we've closed the efficiency gap.

Eh, the type reconciliation should mostly close the gap. (So long as no one ever provides limits for `varchar`/`bpchar`, we'll be able to flatten everything.)

ggevay commented:
Recording here that custom ids in ListCreate might be another thing that we want to erase: https://github.com/MaterializeInc/materialize/pull/32486#discussion_r2089071252

frankmcsherry commented:
This came up in office hours, in the context of a more vigorous version (perhaps a step 4. if this works out) where we unify each of "integer" types, "bytes" types, floats, maps, arrays, records, etc. The goal would be to be able to reduce down to the distinct ways we represent data, and push the interpretation of them back to function. E.g. you want to add two integers, great; should they wrap around at some number of bytes? Put that in your function, rather than in the type definition. We already have `Row` level tags that cause the number of bytes used to be the same for `u8` as for a `u64` containing a `u8` value; no need to maintain all the distinctions, and similar core problems around index planning and such (equating an `i32` with an `i64` or `u64` or `u8` rules out index re-use, when .. they could all have a uniform representation underneath. Perhaps `i` and `u` variants would be different because of their maximum range, but .. we could also just pack everything into a `i64` and too bad (actual preference: variable length encoding; but we might want the tag to hold the sign in order to avoid a regression).

mgree commented:
I think this is a fine idea; the "numeric tower" often has some surprising corner cases, but nothing we can't handle. (For example: if we have an index using `plus_u8`, and someone uses `plus_u16`, can we use the index? How do we prove that our values will fit? How do we debug missed index opportunities?)

The ideal answer is the one were the type system we're using describes our values as precisely as possible: if we're unifying numbers at the `Row`/`Datum` level, we should do so for types as well.

frankmcsherry commented:
I see now that I got tagged a bunch earlier on, and I should have noticed (it's in the pile of 5k+ unread GH messages T.T).

One thought though was that while there was some concern about the "potentially erroring casts", I think in the case of indexes it's probably not that much of a problem. We have the lucky case that if you equate a `u32` and a `u64` for example, then we can go in either direction up or down, as long as we *discard* the failed casts from `u64` to `u32`, rather than error on them. This holds for equality, at least, though 100% things get confusing and imo we should bail out pretty quickly when someone starts to do things like `add_u8` that present as just weird functions. At least, we should bail out until we have a more thorough mastery of that. But I think less risk of something bad happening when one equates differently sized or signed integers: they match exactly when they match in the unbounded integers, I think?

frankmcsherry commented:
> The ideal answer is the one were the type system we're using describes our values as precisely as possible: if we're unifying numbers at the Row/Datum level, we should do so for types as well.

Strong agree here, but also .. happy to let this steep for a bit if we want to mull it or learn a bit first. I'd love to get some related clarity about how we are using types, for example around nullability and whether this is type-level information (imo, it really oughtn't be relied on, due to predicate movement, and types are things you should be able to rely on; said by someone who didn't do great in PL though).

mgree commented:
I'm totally content with not representing nullability in types---we're not tracking it precisely enough to be consistent or useful. But we _are_ using it in, e.g., `MSE::Reduce` to determine how to handle things.

A possible way forward: remove nullability from types, only using it as part of an analysis. A downside is that we will do less optimization (and pay the price of more analysis) in some cases... and we might not be happy with how much we regress.

ggevay reviewed (approved):
Thank you, this looks great!

My only lingering worry is how big the `RelationDesc` refactoring will be. Maybe you could do a quick scan of the (861...) places in the code where `RelationDesc` is mentioned before merging the design doc.

And one last thing: while doing a big overhaul of the type system, could you also take a quick look at `CatalogType` (and maybe even `TypeReference`, `ResolvedDataType`, `RawDataType`, `BuiltinType`, etc.)? An example of a related question that the design doc might want to explicitly answer is what happens with `scalar_type_from_catalog`? It will be turned into `sql_scalar_type_from_catalog`, and we hopefully won't need a `repr_scalar_type_from_catalog`, right? Also, is there any code currently that maps from `ScalarType` to `CatalogType`? (https://github.com/MaterializeInc/database-issues/issues/8059#issuecomment-2370006604 is an example issue where we have a type system bug at the level of `CatalogType`.)

ggevay commented:
One more thing: Among the possible benefits and long-term plans, you could also mention that idea from Frank where all number types would be represented as arbitrary-precision integers, in which case their repr types would be the same, which would open up many mor ecast-removal possiblities.

ggevay reviewed (commented):
Some more questions:

ggevay commented on doc/developer/design/20240522_mir_representation_types.md:
```diff
@@ -0,0 +1,192 @@
+## Success Criteria
+
+1. Plan joins without implicit coercions between equivalent representation types getting in the way.
+2. It should be possible for us to elide noop casts at the MIR level.
```
(Some context: the `invert_casts_on_expr_eq_literal` stuff can't solve cast problems in joins, because a cast on any side of a join equality is bad, so moving a cast to the other side doesn't help. It does work for literal equalities because we need an index only for one side, and when the cast lands on the literal side, then we can just evaluate it by const-folding, so the cast just disappears.)

mgree commented on doc/developer/design/20240522_mir_representation_types.md:
```diff
@@ -0,0 +1,192 @@
+    + `AggregateFunc` is already split; the `expr` crate's `MapAgg` can use `ReprScalarType`.
+    + `UnaryFunc` doesn't take any type arguments.
+    + `BinaryFunc::RangeContainsElem` changes to use `ReprScalarType`.
+    + `VariadicFunc` uses `ScalarType`. We can split it or parameterize it.
```
I'll try to parameterize if I can (but that will still be a non-trivial diff 🥲 my specialty, apparently 🙃).

mgree commented on doc/developer/design/20240522_mir_representation_types.md:
```diff
@@ -0,0 +1,192 @@
+
+## Success Criteria
+
+1. Plan joins without implicit coercions between equivalent representation types getting in the way.
```
I doubt it---if there is such a case, it's a hole in that function! But this is a good test critierion; I'll add it.

frankmcsherry reviewed (approved):
This looks good to me. It looks like there are several tactical questions about the order of next steps, but I think "strategically" this is an important unlock in building abstraction boundaries in to decouple the concerns of the parts. I had some async conversations with @mgree about "optimizer architecture" and that this is one way to think about the same: we are describing something closer to "backend types", though also that's a thing to avoid (we don't want to be coupled to a backend) and that this may be splitting apart "frontend types" and "IR types".
